handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -614,21 +614,34 @@ public final SSLEngineResult wrap(
                 }
             }
 
-            if (dst.remaining() < MAX_ENCRYPTED_PACKET_LENGTH) {
+            int endOffset = offset + length;
+            int srcsLen = 0;
+
+            for (int i = offset; i < endOffset; ++i) {
+                final ByteBuffer src = srcs[i];
+                if (src == null) {
+                    throw new IllegalArgumentException("srcs[" + i + "] is null");
+                }
+                srcsLen += src.remaining();
+            }
+
+            int maxEncryptedLen = calculateOutNetBufSize(srcsLen);
+
+            // If overflow happened just use MAX_PLAINTEXT_LENGTH.
+            if (maxEncryptedLen < 0 || MAX_ENCRYPTED_PACKET_LENGTH < maxEncryptedLen) {
```

Comment:
@Scottmitch @trustin @nmittler in theory this could fail to detect overflow if it overflows enough to again be >= 0. Not sure if we need to care enough about it to justify moving the check inside the for loop and so have more branches. WDYT ?